cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -736,35 +736,102 @@ private module ImplCommon {
     else result instanceof LocalCallContextAny
   }
 
+  /**
+   * A node from which flow can return to the caller. This is either a regular
+   * `ReturnNode` or a `PostUpdateNode` corresponding to the value of a parameter.
+   */
+  class ReturnNodeExt extends Node {
+    ReturnNodeExt() {
+      this instanceof ReturnNode or
+      parameterValueFlowsToUpdate(_, this)
+    }
+
+    /** Gets the kind of this returned value. */
+    ReturnKindExt getKind() {
+      result = TValueReturn(this.(ReturnNode).getKind())
+      or
+      exists(ParameterNode p, int pos |
+        parameterValueFlowsToUpdate(p, this) and
+        p.isParameterOf(_, pos) and
+        result = TParamUpdate(pos)
+      )
+    }
+  }
+
+  private newtype TReturnKindExt =
```

Comment:
I think we should cache this, like we cache e.g. `TReturnPosition`.

Reply:
I haven't seen any need for this as it's a fairly light-weight predicate - indeed perhaps `TReturnPosition` shouldn't be cached either.